yt/frontends/nc4_cm1/data_structures.py

```diff
@@ -0,0 +1,233 @@
+        # assumed to be in code units; domain_left_edge and domain_right_edge
+        # will be converted to YTArray automatically at a later time.
+        # This includes the cosmological parameters.
+        #
+        #   self.unique_identifier      <= unique identifier for the dataset
+        #                                  being read (e.g., UUID or ST_CTIME)
+        self.unique_identifier = int(os.stat(self.parameter_filename)[stat.ST_CTIME])
+        self.parameters = {}  # code-specific items
+        with self._handle.open_ds() as ds:
+            dims = [ds.dimensions[i].size for i in ["xh", "yh", "zh"]]
+            coords = {i: ds.variables[i][:] for i in ["xh", "yh", "zh"]}
+
+            # TO DO: eneralize this to be coordiante variable name agnostic in order to
+            # make useful for WRF or climate data. For now, we're hard coding for CM1
+            # specifically and have named the classes appropriately
+            xh, yh, zh = [ds.variables[i][:] for i in ["xh", "yh", "zh"]]
+            self.domain_left_edge = np.array(
+                [xh.min(), yh.min(), zh.min()], dtype="float64"
+            )
+            self.domain_right_edge = np.array(
+                [xh.max(), yh.max(), zh.max()], dtype="float64"
+            )
+
+            # loop over the variable names in the netCDF file, record only those on the
+            # "zh","yh","xh" grid.
+            varnames = []
+            for key in ds.variables.keys():
+                vardims = ds.variables[key].dimensions
+                if all(x in vardims for x in ["time", "zh", "yh", "xh"]) is True:
+                    varnames.append(key)
+            self.parameters["variable_names"] = varnames
+            self.parameters["lofs_version"] = ds.cm1_lofs_version
+            self.parameters["is_uniform"] = ds.uniform_mesh
+            self.current_time = ds.variables["time"][:][0]
+
+            # record the dimension metadata
+            dim_info = OrderedDict()
+            for dim, meta in ds.dimensions.items():
+                dim_info[dim] = meta.size
+            self.parameters["dimensions"] = dim_info
+
+        self.parameters["coords"] = coords
+
+        self.dimensionality = 3
+        self.domain_dimensions = np.array(dims, dtype="int64")
+        self.periodicity = (False, False, False)
+        self.parameters["time"] = self.current_time
+
+        # Set cosmological information to zero for non-cosmological.
+        self.cosmological_simulation = 0.0
+        self.current_redshift = 0.0
+        self.omega_lambda = 0.0
+        self.omega_matter = 0.0
+        self.hubble_constant = 0.0
+
+    @classmethod
+    def _is_valid(self, *args, **kwargs):
+        # This accepts a filename or a set of arguments and returns True or
+        # False depending on if the file is of the type requested.
+
+        warn_netcdf(args[0])
+        try:
+            from netCDF4 import Dataset
```

Comment:
I think it could be... this follows the `exodus_ii` `is_valid()` but I'm not sure if there was a reason for importing netcdf directly there either.

Reply:
oh, sorry, was conflating two issues. You just meant the import. Yes, should definitely be able to use the `_netCDF4` import. But I've been confused about why the `is_valid` (here and in exodus_ii) did not use the `NetCDF4FileHandler` that is already imported. After more digging, the reason for **that** seems to be the need for the `keepweakref` argument, since `NetCDF4FileHandler.open_ds` uses hardcoded arguments for the call to the netcdf dataset. So I'm inclined to add `**kwargs` to `NetCDF4FileHandler.open_ds` so that we can simply pass along the `keepweakref` using the same `NetCDF4FileHandler` used in the rest of the front end.

Reply:
latest commit fixes this: uses the NetCDF4FileHandler and no longer imports from netCDF4 directly.